numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -355,6 +401,76 @@ init_comparison(PyObject *umath)
 }
 
 
+static NPY_CASTING
+string_addition_resolve_descriptors(
+        PyArrayMethodObject *NPY_UNUSED(self),
+        PyArray_DTypeMeta *NPY_UNUSED(dtypes[3]),
+        PyArray_Descr *given_descrs[3],
+        PyArray_Descr *loop_descrs[3],
+        npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
+    Py_INCREF(given_descrs[1]);
+    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
```

Comment:
Adding this was required to not get nonsense results and it would have been good to add a test that would fail with this missing:  I.e. a byte-swapped unicode string input or correctly sized output.

Reply:
I added this test:

```python3
@pytest.mark.parametrize("order", [">", "<"])
def test_addition_unicode_inverse_byte_order(order):
    element = 'abcd'
    arr1 = np.array([element], dtype=f"{order}U4")
    arr2 = np.array([element], dtype=f"{order}U4")
    result = arr1 + arr2
    assert result == 2*element
```

It oddly fails when with `NPY_DT_CALL_ensure_canonical` there, but it succeeds without it. Is there anything wrong with the test?